### PR TITLE
Conf store/feature/conf store

### PR DESCRIPTION
**Pull request by suryakumar1024**
**Changes**
created new package for kvstore and renamed to kv_store -> existing kvstore directory remains the same
Added confStore and FileStorage with api's. Currently works with jsonfilestore
Added load, get, set and save for json filestore.

### CODE REVIEW

File: py-utils/src/utils/conf_store/file_storage.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import os
+from src.utils.kv_store import KvStorage
+
+class FileStorage(KvStorage):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/file_storage.py:4]: This class should be part of kv_store

File: py-utils/src/utils/conf_store/kvstore_factory.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import os
+from src.utils.conf_store.storage import JsonStorage, YamlStorage
+
+class KvStoreFactory:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/kvstore_factory.py:4]: This class should be part of kv_store

File: py-utils/src/utils/conf_store/storage/json_storage.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import json
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/storage/json_storage.py:1]: This class should be part of kv_store

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from src.utils.schema.payload import Payload
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Updated in current patch. Kept the same code as ConfCache inside conf_store

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                if key is not None:
+                    key_sequence = "".join(str([e]) for e in key.split('.'))
+                    return eval(str(self._configuration_payload[index])+key_sequence)
+                else:
+                    return self._configuration_payload[index]
+        except KeyError:
+            print('Value for given key index not found')
+            return default_value
+            
+    def load(self, index, value):
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+from src.utils.schema.payload import Payload
+
+class ConfCache:
+
+    def __init__(self):
+       self._configuration_payload = {}
+       # need to be discussed
+       self.payload = Payload
+
+    def get(self, index, key=None, default_value = None):
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+from src.utils.schema.payload import Payload
+
+class ConfCache:
+
+    def __init__(self):
+       self._configuration_payload = {}
+       # need to be discussed
+       self.payload = Payload
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: updated in current patch. Copied payload as in the name of ConfCache

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+        try:
+            if index in self._configuration_payload:
+                if key is not None:
+                    key_sequence = "".join(str([e]) for e in key.split('.'))
+                    return eval(str(self._configuration_payload[index])+key_sequence)
+                else:
+                    return self._configuration_payload[index]
+        except KeyError:
+            print('Value for given key index not found')
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Removed this code. updated with new code in current patch

File: py-utils/src/utils/conf_store/conf_storage.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        # List of keys and values
+        return self._conf_cache.setter(index, keyList, valueList)
+
+    def save(self, index):
+        data = self.get(index)
+        self._store.dump(data)
+
+    def backup(self):
+        pass
+    def copy(self, index1, index2):
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_storage.py]: updated in current patch

File: py-utils/src/utils/conf_store/conf_type.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+from __future__ import absolute_import
+
+class ConfType():
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_type.py]: updated in current patch. removed ConfType class from confStore

File: py-utils/src/utils/conf_store/file_storage.py (hunk 1 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/file_storage.py]: updated code. current implementation is different from previous code architecture

File: py-utils/src/utils/conf_store/kvstore_factory.py (hunk 2 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/kvstore_factory.py]: updated in current patch. current implementation doesn't have this code

File: py-utils/src/utils/conf_store/storage/json_storage.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/storage/json_storage.py]: updated in current patch. current implementation doesn't have this code file

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+
+class ConfCache:
+    """
+    ConfCache implements in-memory configuration management in specified format
+    """
+
+    def __init__(self, kv_store_instance):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cache.py:24]:
```
def __init__(self, kv_store: KvStore):

File: py-utils/src/utils/conf_store/conf_store.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    def __init__(self):
+        """
+        kvstore will be initialized at the time of load
+        """
+        self.kvstore = ''
+        self._payload = {}
+
+    def load(self, index, kvStore, force=False, callback=None) -> None:
+        """
+        Args:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:34]:
```
parameters:
index:     Identifier for the config loaded from the KV Store
kv_store:  KV Store (Conf Backend)
overwrite: When False, it throws exception if index already exists. 
           Default: False 
callback:  Callback for the config changes in the KV Store
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    ConfStore implements configuration management on top of KvStore
+    """
+    def __init__(self):
+        """
+        kvstore will be initialized at the time of load
+        """
+        self.kvstore = ''
+        self._payload = {}
+
+    def load(self, index, kvStore, force=False, callback=None) -> None:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:32]: kvStore is incorrect variable name. We use underscore naming conventions. Please use consistent coding conventions.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Comment by suryakumar1024** [comment]
Have raised new PR with new Architectural level changes.  So closing this

**Pull request closed without merging**